llama_index/llms/llm_prompt_mixin.py

```diff
@@ -0,0 +1,254 @@
+from collections import ChainMap
+from typing import (
+    Any,
+    Callable,
+    List,
+    Optional,
+)
+
+from llama_index.bridge.pydantic import BaseModel, Field
+from llama_index.callbacks import CBEventType, EventPayload
+from llama_index.llms.base import LLM
+from llama_index.llms.types import (
+    ChatMessage,
+    ChatResponseAsyncGen,
+    ChatResponseGen,
+    CompletionResponseAsyncGen,
+    CompletionResponseGen,
+    MessageRole,
+)
+from llama_index.prompts import BasePromptTemplate, PromptTemplate
+from llama_index.types import PydanticProgramMode, TokenAsyncGen, TokenGen
+
+
+def stream_completion_response_to_tokens(
+    completion_response_gen: CompletionResponseGen,
+) -> TokenGen:
+    """Convert a stream completion response to a stream of tokens."""
+
+    def gen() -> TokenGen:
+        for response in completion_response_gen:
+            yield response.delta or ""
+
+    return gen()
+
+
+def stream_chat_response_to_tokens(
+    chat_response_gen: ChatResponseGen,
+) -> TokenGen:
+    """Convert a stream completion response to a stream of tokens."""
+
+    def gen() -> TokenGen:
+        for response in chat_response_gen:
+            yield response.delta or ""
+
+    return gen()
+
+
+async def astream_completion_response_to_tokens(
+    completion_response_gen: CompletionResponseAsyncGen,
+) -> TokenAsyncGen:
+    """Convert a stream completion response to a stream of tokens."""
+
+    async def gen() -> TokenAsyncGen:
+        async for response in completion_response_gen:
+            yield response.delta or ""
+
+    return gen()
+
+
+async def astream_chat_response_to_tokens(
+    chat_response_gen: ChatResponseAsyncGen,
+) -> TokenAsyncGen:
+    """Convert a stream completion response to a stream of tokens."""
+
+    async def gen() -> TokenAsyncGen:
+        async for response in chat_response_gen:
+            yield response.delta or ""
+
+    return gen()
+
+
+class LLMPromptMixin(LLM):
+    system_prompt: Optional[str] = Field(description="System prompt for LLM calls.")
+    messages_to_prompt: Optional[Callable[[List[ChatMessage]], str]] = Field(
+        description="Function to convert a list of messages to an LLM prompt.",
+        exclude=True,
+    )
+    completion_to_prompt: Optional[Callable[[str], str]] = Field(
+        description="Function to convert a completion to an LLM prompt.", exclude=True
+    )
+    pydantic_program_mode: PydanticProgramMode = PydanticProgramMode.DEFAULT
+
+    # deprecated
```

Comment:
if this is a new abstraction, do we need this? I presume even in moving to this new abstraction you want to maintain old UX for a little bit?

Reply:
Wanted to keep it backwards compatible with the old LLMPredictor -- so needed to keep this (for now) for feature parity